README.md
# skeletonize

[![Build Status](https://github.com/okaneco/skeletonize/workflows/Rust%20CI/badge.svg)](https://github.com/okaneco/skeletonize/)
[![Crates.io](https://img.shields.io/crates/v/skeletonize.svg)](https://crates.io/crates/skeletonize)
[![Docs.rs](https://docs.rs/skeletonize/badge.svg)](https://docs.rs/skeletonize)

!["skeletonize" original text and line thinned text](gfx/skeletonize-title.png)

A line thinning library for binary images, including edge detection and
threshold functions for preprocessing images into binary images.

The goal of line thinning is to remove excess pixels from the image until the
lines present are one pixel wide, resembling a "skeleton" of the original
pattern. Thinning is useful for removing noise from images which have had
image processing filters applied to them such as edge detection. Line thinning
is similar to erosion, another morphological operator.

The thinning algorithms are based on the papers *Zhang & Suen, 1984* and
*Chen & Hsu, 1988*. See [Reference](#reference).

This crate requires the input to be a type from the `image` crate. To use this
crate, add the following to your `Cargo.toml`.

```toml
[dependencies.skeletonize]
version = "0.2"
```

#### Features

- 2 line thinning algorithms
- support for black or white foreground color
- Sobel operator edge detection
- thresholding for binarization (turning an image into only black and
white pixels)

The example [`skeletonize.rs`](examples/skeletonize.rs) is a command line
program available for download as a binary executable from the repository
[`Releases`][releases] page.

## Thinning Algorithms Comparison

*Images should be viewed at 100% magnification to avoid scaling artifacts.*

![Original image and line thinning processed images](gfx/chenhsu-append.png)
![Comparison of Standard and Modified algorithms](gfx/chenhsu-standard-modified.gif)  
*First image, left to right: Figures 7(b) and 7(c) from Chen & Hsu, 1988,
`Standard` algorithm result, `Modified` algorithm result.*  
*Second image: Animated comparison of the `Standard` and `Modified` marking
methods.*

The *Chen & Hsu* `Modified` algorithm tends to produce better line connectivity,
less noise, and more consistent single pixel width lines than the *Zhang & Suen*
`Standard` algorithm.

The original image can be found in the `gfx` folder entitled `chenhsu.png`. It
was thresholded at `0.3` to convert it into a binary image before line thinning.
The example program produced the image results with the following arguments. The
`--method|-m` option allows for selecting the standard or modified pixel marking
algorithm, `--threshold|-t` is the gray level threshold.

```bash
-i chenhsu.png -t 0.3 -m s -o chenhsu-standard.png
-i chenhsu.png -t 0.3 -m m -o chenhsu-modified.png
```

## [`skeletonize.rs`](examples/skeletonize.rs)

The `skeletonize` example program exposes library functions as a command line
application. It can be run with the following command or by invoking it directly
after downloading from [`Releases`][releases]/building it yourself.

```bash
cargo r --release --example skeletonize -- [args]
```
### Examples

The next three examples use this image as the input. All examples include the
equivalent library code.

![Original image](gfx/rustacean.png)

---

Perform edge detection with `sobel4` (4-way edge detection) and line thinning,
threshold the edge detection filter to `0.3`.

```bash
cargo r --release --example skeletonize -- -i rustacean.png -e sobel4 -t 0.3
```

```rust
let mut filtered = sobel4::<foreground::Black>(&img, Some(0.3))?;
thin_image_edges::<foreground::Black>(&mut filtered, method, None)?;
```

![Edge detected crab](gfx/edgedetect-t03.png)

---

Perform edge detection with no line thinning, threshold the edge detection
filter to `0.3`, and set the `--foreground|-f` color to white.

```bash
-i rustacean.png -e sobel4 -t 0.3 --no-thin -f white
```

```rust
let filtered = sobel4::<foreground::White>(&img, Some(0.3))?;
```

![Edge detected crab with no thinning](gfx/edgedetect-no-thin.png)

---

Return the grayscale edge detection image by omitting the `--threshold|-t` and
using `--no-thin`. Aliases are used for `sobel4` and `white`.

```bash
-i rustacean.png -e s4 --no-thin -f w
```

```rust
let filtered = sobel4::<foreground::White>(&img, None))?;
```

![Edge detected crab with no thinning](gfx/edgedetect-no-thin-no-threshold.png)

---

The title image was created with the following arguments.

```bash
-i skeletonize.png -t 0.3
```

```rust
skeletonize::threshold(&mut img, 0.3)?;
thin_image_edges::<foreground::Black>(&mut img, method, None)?;
```

<a name="reference"></a>
## Reference

*Zhang, T. Y. & Suen, C. Y. (1984). A fast parallel algorithm for thinning
digital patterns. Commun. ACM 27, 3 (March 1984), 236–239.
[DOI:10.1145/357994.358023](https://doi.org/10.1145/357994.358023)*

*Chen, Yung-Sheng & Hsu, Wen-Hsing. (1988). A modified fast parallel algorithm
for thinning digital patterns. Pattern Recognition Letters. 7. 99-106.
[DOI:10.1016/0167-8655(88)90124-9](https://doi.org/10.1016/0167-8655(88)90124-9)*

## License

This crate is licensed under either
- the [MIT License](LICENSE-MIT), or
- the [Apache License (Version 2.0)](LICENSE-APACHE)

at your option.

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall be
dual licensed as above, without any additional terms or conditions.


[releases]: https://github.com/okaneco/skeletonize/releases


CHANGELOG.md
# `skeletonize` changelog

## Version 0.2.0 - 2023-01-14
Bumped `image` dependency to `0.24`.

[#2][2] - Upgrade `image` to `0.24.5`, prepare new release

## Version 0.1.0 - 2021-05-01
- Initial release

[2]: https://github.com/okaneco/skeletonize/pull/2


src/error.rs
//! Image thinning error enums.

/// Error for edge thinning and edge detection.
#[derive(Copy, Clone, Debug)]
pub enum SkeletonizeError {
    /// Error converting an image to a grayscale image.
    LumaConversion(LumaConversionErrorKind),
    /// The edge thinning algorithm reached the maximum amount of iterations.
    MaxThinningIterations,
}

/// Errors that occur when attempting to convert an image to grayscale.
#[derive(Copy, Clone, Debug)]
pub enum LumaConversionErrorKind {
    /// Error converting an image into grayscale in an edge thinning algorithm.
    ImageThinningLuma,
    /// Error converting an image to grayscale in an edge detection function.
    SobelLuma,
    /// Error converting an image into a mutable grayscale image view in an edge
    /// detection function.
    SobelMutableLuma,
    /// Error converting an image into a mutable grayscale image view for
    /// thresholding.
    ThresholdMutableLuma,
}

impl core::fmt::Display for LumaConversionErrorKind {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::ImageThinningLuma => {
                write!(f, "Could not create a grayscale image in image thinning")
            }
            Self::SobelLuma => write!(f, "Could not create a grayscale image in edge detection"),
            Self::SobelMutableLuma => write!(
                f,
                "Could not create a mutable grayscale image view in edge detection"
            ),
            Self::ThresholdMutableLuma => write!(
                f,
                "Could not create a mutable grayscale image view for thresholding"
            ),
        }
    }
}

impl core::fmt::Display for SkeletonizeError {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::LumaConversion(err) => write!(f, "{err}"),
            Self::MaxThinningIterations => {
                write!(f, "Maximum iteration count reached in thinning algorithm")
            }
        }
    }
}

impl std::error::Error for SkeletonizeError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            Self::LumaConversion(_) | Self::MaxThinningIterations => None,
        }
    }
}


src/thinning.rs
//! Functions for performing image thinning.

use image::GenericImage;

use crate::error::{LumaConversionErrorKind, SkeletonizeError};
use crate::neighbors::get_neighbor_info;
use crate::{Edge, ForegroundColor, MarkingMethod};

/// Perform image thinning on a binarized image `img` using one of the methods
/// in [`MarkingMethod`](crate::MarkingMethod). Returns the number of iterations
/// needed for thinning on successful completion.
///
/// `iterations` is an optional parameter set to `u32::MAX` if `None`.
#[allow(clippy::collapsible_else_if)]
#[allow(clippy::nonminimal_bool)]
pub fn thin_image_edges<F: ForegroundColor>(
    img: &mut image::DynamicImage,
    method: MarkingMethod,
    iterations: Option<u32>,
) -> Result<u32, SkeletonizeError> {
    let mut pixels_to_remove = Vec::new();
    let mut phase_one = true;
    let iterations = iterations.unwrap_or(u32::MAX);

    for iters in 0..iterations {
        let luma_img = img.as_luma8().ok_or(SkeletonizeError::LumaConversion(
            LumaConversionErrorKind::ImageThinningLuma,
        ))?;
        let (width, height) = luma_img.dimensions();

        // Mark pixels to remove
        for (x, y, p) in luma_img.enumerate_pixels() {
            if *p == image::Luma([F::BACKGROUND_COLOR]) {
                continue;
            }

            let info = get_neighbor_info::<F>(luma_img, width, height, x, y);
            let [p2, p3, p4, p5, p6, p7, p8, p9] = info.edge_status;

            match method {
                MarkingMethod::Standard => {
                    // Zhang and Suen, 1984

                    // Don't mark if we don't have 8 neighbors and 2..=6 aren't filled
                    if !(2..=6).contains(&info.filled) || info.neighbors != 8 {
                        continue;
                    }

                    // Count the number of times an edge transitions from empty to filled
                    let mut transitions = 0;
                    for pair in [p2, p3, p4, p5, p6, p7, p8, p9, p2].windows(2) {
                        if pair[0] == Edge::Empty && pair[1] == Edge::Filled {
                            transitions += 1;
                        }
                    }

                    if transitions != 1 {
                        continue;
                    }

                    if phase_one {
                        if (p2 == Edge::Empty || p4 == Edge::Empty || p6 == Edge::Empty)
                            && (p4 == Edge::Empty || p6 == Edge::Empty || p8 == Edge::Empty)
                        {
                            pixels_to_remove.push((x, y));
                        }
                    } else {
                        if (p2 == Edge::Empty || p4 == Edge::Empty || p8 == Edge::Empty)
                            && (p2 == Edge::Empty || p6 == Edge::Empty || p8 == Edge::Empty)
                        {
                            pixels_to_remove.push((x, y));
                        }
                    }
                }
                MarkingMethod::Modified => {
                    // Chen and Hsu, 1988

                    if !(2..=7).contains(&info.filled) || info.neighbors != 8 {
                        continue;
                    }

                    let mut transitions = 0;
                    for pair in [p2, p3, p4, p5, p6, p7, p8, p9, p2].windows(2) {
                        if pair[0] == Edge::Empty && pair[1] == Edge::Filled {
                            transitions += 1;
                        }
                    }

                    if !(transitions == 1 || transitions == 2) {
                        continue;
                    }

                    if phase_one {
                        if transitions == 1 {
                            if (p2 == Edge::Empty || p4 == Edge::Empty || p6 == Edge::Empty)
                                && (p4 == Edge::Empty || p6 == Edge::Empty || p8 == Edge::Empty)
                            {
                                pixels_to_remove.push((x, y));
                            }
                        } else {
                            if ((p2 == Edge::Filled && p4 == Edge::Filled)
                                && (p6 == Edge::Empty && p7 == Edge::Empty && p8 == Edge::Empty))
                                || ((p4 == Edge::Filled && p6 == Edge::Filled)
                                    && (p2 == Edge::Empty
                                        && p8 == Edge::Empty
                                        && p9 == Edge::Empty))
                            {
                                pixels_to_remove.push((x, y));
                            }
                        }
                    } else {
                        if transitions == 1 {
                            if (p2 == Edge::Empty || p4 == Edge::Empty || p8 == Edge::Empty)
                                && (p2 == Edge::Empty || p6 == Edge::Empty || p8 == Edge::Empty)
                            {
                                pixels_to_remove.push((x, y));
                            }
                        } else {
                            if ((p2 == Edge::Filled && p8 == Edge::Filled)
                                && (p4 == Edge::Empty && p5 == Edge::Empty && p6 == Edge::Empty))
                                || ((p6 == Edge::Filled && p8 == Edge::Filled)
                                    && (p2 == Edge::Empty
                                        && p3 == Edge::Empty
                                        && p4 == Edge::Empty))
                            {
                                pixels_to_remove.push((x, y));
                            }
                        }
                    }
                }
            }
        }

        phase_one = !phase_one;

        // Replace marked pixels with background color to thin the edges
        for &(x, y) in &pixels_to_remove {
            img.put_pixel(x, y, image::Rgba([F::BACKGROUND_COLOR; 4]));
        }

        if pixels_to_remove.is_empty() {
            return Ok(iters);
        }

        pixels_to_remove.clear();
    }

    Err(SkeletonizeError::MaxThinningIterations)
}


src/neighbors.rs
//! Struct and utilities for calculating the status of neighboring pixels.

use crate::{Edge, ForegroundColor};

/// Struct with information describing the surrounding pixels.
pub struct NeighborInfo {
    /// The number of neighbor pixels with non-background color.
    pub filled: u8,
    /// The number of surrounding pixels.
    pub neighbors: u8,
    /// An array containing the [status](crate::Edge) of the neighboring pixels.
    pub edge_status: [Edge; 8],
}

impl NeighborInfo {
    /// Calculate and return the number of transitions from
    /// [`Edge::Empty`](crate::Edge::Empty) to
    /// [`Edge::Filled`](crate::Edge::Filled).
    pub fn transitions(&self) -> u8 {
        let [p2, p3, p4, p5, p6, p7, p8, p9] = self.edge_status;

        let mut transitions = 0;
        for pair in [p2, p3, p4, p5, p6, p7, p8, p9, p2].windows(2) {
            if pair[0] == Edge::Empty && pair[1] == Edge::Filled {
                transitions += 1;
            }
        }

        transitions
    }
}

/// Calculate and return a [`NeighborInfo`](crate::neighbors::NeighborInfo)
/// struct which contains the number of occupied neighbor pixels, number of
/// neighbor pixels, and the [edge status](crate::Edge) of the neighboring
/// pixels.
pub fn get_neighbor_info<F: ForegroundColor>(
    img: &image::GrayImage,
    width: u32,
    height: u32,
    x: u32,
    y: u32,
) -> NeighborInfo {
    let mut filled = 0;
    let mut neighbors = 0;

    let p9 = if y > 0 && x > 0 {
        neighbors += 1;
        if img.get_pixel(x - 1, y - 1)[0] != F::BACKGROUND_COLOR {
            filled += 1;
            Edge::Filled
        } else {
            Edge::Empty
        }
    } else {
        Edge::DoesNotExist
    };
    let p2 = if y > 0 {
        neighbors += 1;
        if img.get_pixel(x, y - 1)[0] != F::BACKGROUND_COLOR {
            filled += 1;
            Edge::Filled
        } else {
            Edge::Empty
        }
    } else {
        Edge::DoesNotExist
    };
    let p3 = if y > 0 && x < u32::MAX && x + 1 < width {
        neighbors += 1;
        if img.get_pixel(x + 1, y - 1)[0] != F::BACKGROUND_COLOR {
            filled += 1;
            Edge::Filled
        } else {
            Edge::Empty
        }
    } else {
        Edge::DoesNotExist
    };
    let p8 = if x > 0 {
        neighbors += 1;
        if img.get_pixel(x - 1, y)[0] != F::BACKGROUND_COLOR {
            filled += 1;
            Edge::Filled
        } else {
            Edge::Empty
        }
    } else {
        Edge::DoesNotExist
    };
    let p4 = if x < u32::MAX && x + 1 < width {
        neighbors += 1;
        if img.get_pixel(x + 1, y)[0] != F::BACKGROUND_COLOR {
            filled += 1;
            Edge::Filled
        } else {
            Edge::Empty
        }
    } else {
        Edge::DoesNotExist
    };
    let p7 = if x > 0 && y < u32::MAX && y + 1 < height {
        neighbors += 1;
        if img.get_pixel(x - 1, y + 1)[0] != F::BACKGROUND_COLOR {
            filled += 1;
            Edge::Filled
        } else {
            Edge::Empty
        }
    } else {
        Edge::DoesNotExist
    };
    let p6 = if y < u32::MAX && y + 1 < height {
        neighbors += 1;
        if img.get_pixel(x, y + 1)[0] != F::BACKGROUND_COLOR {
            filled += 1;
            Edge::Filled
        } else {
            Edge::Empty
        }
    } else {
        Edge::DoesNotExist
    };
    let p5 = if x < u32::MAX && x + 1 < width && y < u32::MAX && y + 1 < height {
        neighbors += 1;
        if img.get_pixel(x + 1, y + 1)[0] != F::BACKGROUND_COLOR {
            filled += 1;
            Edge::Filled
        } else {
            Edge::Empty
        }
    } else {
        Edge::DoesNotExist
    };

    NeighborInfo {
        filled,
        neighbors,
        edge_status: [p2, p3, p4, p5, p6, p7, p8, p9],
    }
}


src/edge_detection.rs
//! Edge detection algorithms for preprocessing images.

use crate::error::{LumaConversionErrorKind, SkeletonizeError};
use crate::ForegroundColor;

/// Sobel vertical `North` gradient operator.
#[rustfmt::skip]
pub const SOBEL_NORTH: [f32; 9] = [
    1.0, 2.0, 1.0,
    0.0, 0.0, 0.0,
    -1.0, -2.0, -1.0,
];
/// Sobel vertical `South` gradient operator.
#[rustfmt::skip]
pub const SOBEL_SOUTH: [f32; 9] = [
    -1.0, -2.0, -1.0,
    0.0, 0.0, 0.0,
    1.0, 2.0, 1.0,
];
/// Sobel horizontal `East` gradient operator.
#[rustfmt::skip]
pub const SOBEL_EAST: [f32; 9] = [
    -1.0, 0.0, 1.0,
    -2.0, 0.0, 2.0,
    -1.0, 0.0, 1.0,
];
/// Sobel horizontal `West` gradient operator.
#[rustfmt::skip]
pub const SOBEL_WEST: [f32; 9] = [
    1.0, 0.0, -1.0,
    2.0, 0.0, -2.0,
    1.0, 0.0, -1.0,
];

/// Detect edges in an image using [`SOBEL_EAST`](SOBEL_EAST) and
/// [`SOBEL_NORTH`](SOBEL_NORTH) gradient operators.
/// The image should not have transparency.
///
/// `threshold` is an optional parameter between 0.0 and 1.0 which is used to
/// binarize the image. Pixels below that `Luma` threshold will be converted
/// to the background color.
pub fn sobel<F: ForegroundColor>(
    img: &image::DynamicImage,
    threshold: Option<f32>,
) -> Result<image::DynamicImage, SkeletonizeError> {
    let mut filter_up = img.filter3x3(&SOBEL_NORTH);
    let filtered_right = img.filter3x3(&SOBEL_EAST);
    let mutable_error = SkeletonizeError::LumaConversion(LumaConversionErrorKind::SobelMutableLuma);
    let immutable_error = SkeletonizeError::LumaConversion(LumaConversionErrorKind::SobelLuma);

    let iter_down = filter_up.as_mut_luma8().ok_or(mutable_error)?.iter_mut();
    let iter_right = filtered_right.as_luma8().ok_or(immutable_error)?.iter();

    for (g_down, g_right) in iter_down.zip(iter_right) {
        let res = (f32::from(*g_down) / 255.0).hypot(f32::from(*g_right) / 255.0);

        if let Some(threshold) = threshold {
            *g_down = if res < threshold {
                F::BACKGROUND_COLOR
            } else {
                !F::BACKGROUND_COLOR
            }
        } else {
            *g_down = (res * 255.0).round() as u8;
        }
    }

    // If ForegroundColor is Black and threshold None, edges would stay white
    // so we need to invert the result before returning it.
    if threshold.is_none() && F::BACKGROUND_COLOR == 255 {
        filter_up.invert()
    }

    Ok(filter_up)
}

/// Detect edges in an image using four Sobel gradient operators:
/// [`SOBEL_NORTH`](SOBEL_NORTH), [`SOBEL_SOUTH`](SOBEL_SOUTH),
/// [`SOBEL_EAST`](SOBEL_EAST), and [`SOBEL_WEST`](SOBEL_WEST).
/// The image should not have transparency.
///
/// `threshold` is an optional parameter between 0.0 and 1.0 which is used to
/// binarize the image. Pixels below that `Luma` threshold will be converted
/// to the background color.
pub fn sobel4<F: ForegroundColor>(
    img: &image::DynamicImage,
    threshold: Option<f32>,
) -> Result<image::DynamicImage, SkeletonizeError> {
    let mut filter_up = img.filter3x3(&SOBEL_NORTH);
    let filter_down = img.filter3x3(&SOBEL_SOUTH);
    let filter_right = img.filter3x3(&SOBEL_EAST);
    let filter_left = img.filter3x3(&SOBEL_WEST);

    let mutable_error = SkeletonizeError::LumaConversion(LumaConversionErrorKind::SobelMutableLuma);
    let immutable_error = SkeletonizeError::LumaConversion(LumaConversionErrorKind::SobelLuma);

    let iter_up = filter_up.as_mut_luma8().ok_or(mutable_error)?.iter_mut();
    let iter_down = filter_down.as_luma8().ok_or(immutable_error)?.iter();
    let iter_right = filter_right.as_luma8().ok_or(immutable_error)?.iter();
    let iter_left = filter_left.as_luma8().ok_or(immutable_error)?.iter();

    for (((g_up, g_down), g_left), g_right) in iter_up.zip(iter_down).zip(iter_right).zip(iter_left)
    {
        let vertical = (f32::from(*g_up) - f32::from(*g_down)) / 255.0;
        let horizontal = (f32::from(*g_right) - f32::from(*g_left)) / 255.0;
        let res = vertical.hypot(horizontal);

        if let Some(threshold) = threshold {
            *g_up = if res < threshold {
                F::BACKGROUND_COLOR
            } else {
                !F::BACKGROUND_COLOR
            }
        } else {
            *g_up = (res * 255.0).round() as u8;
        }
    }

    // If ForegroundColor is Black and threshold None, edges would stay white
    // so we need to invert the result before returning it.
    if threshold.is_none() && F::BACKGROUND_COLOR == 255 {
        filter_up.invert()
    }

    Ok(filter_up)
}


examples/skeletonize.rs
use skeletonize::edge_detection::{sobel, sobel4};
use skeletonize::{foreground, thin_image_edges, MarkingMethod};
use structopt::StructOpt;

fn main() {
    if let Err(e) = try_main() {
        eprintln!("{e}");
        std::process::exit(1);
    }
}

fn try_main() -> Result<(), Box<dyn std::error::Error>> {
    let mut opt = Opt::from_args();

    // Open image and initialize output filename
    let img = image::DynamicImage::ImageLuma8(image::open(&opt.input)?.to_luma8());
    let output = if let Some(output) = opt.output {
        output
    } else {
        generate_filename(&opt.input)?.into()
    };

    // Process the option strings for matching
    opt.edge.make_ascii_lowercase();
    opt.foreground.make_ascii_lowercase();
    opt.method.make_ascii_lowercase();

    let edge = match opt.edge.as_str() {
        "sobel" | "s" => EdgeDetection::Sobel,
        "sobel4" | "s4" => EdgeDetection::Sobel4,
        "" => EdgeDetection::None,
        _ => return Err("Edge detection must be `sobel`/`s` or `sobel4`/`s4`".into()),
    };
    let foreground = match opt.foreground.as_str() {
        "black" | "b" => Fg::Black,
        "white" | "w" => Fg::White,
        _ => return Err("Foreground color must be `black`/`b` or `white`/`w`".into()),
    };
    let method = match opt.method.as_str() {
        "modified" | "m" => MarkingMethod::Modified,
        "standard" | "s" => MarkingMethod::Standard,
        _ => return Err("Method must be `standard`/`s` or `modified`/`m`".into()),
    };

    // Perform edge detection if one of the Sobel options is passed
    let mut filtered = match edge {
        EdgeDetection::Sobel => match foreground {
            Fg::Black => sobel::<foreground::Black>(&img, opt.threshold)?,
            Fg::White => sobel::<foreground::White>(&img, opt.threshold)?,
        },
        EdgeDetection::Sobel4 => match foreground {
            Fg::Black => sobel4::<foreground::Black>(&img, opt.threshold)?,
            Fg::White => sobel4::<foreground::White>(&img, opt.threshold)?,
        },
        EdgeDetection::None => {
            let mut filtered = img;
            if let Some(t) = opt.threshold {
                skeletonize::threshold(&mut filtered, t)?;
            }
            filtered
        }
    };

    // Skip thinning if `no-thin` flag is passed
    if !opt.no_thin {
        match foreground {
            Fg::Black => {
                thin_image_edges::<foreground::Black>(&mut filtered, method, None)?;
            }
            Fg::White => {
                thin_image_edges::<foreground::White>(&mut filtered, method, None)?;
            }
        }
    }

    Ok(filtered.save(output)?)
}

enum EdgeDetection {
    Sobel,
    Sobel4,
    None,
}

enum Fg {
    Black,
    White,
}

#[derive(StructOpt, Debug)]
#[structopt(name = "skeletonize", about = "Image edge thinning utility")]
pub struct Opt {
    /// Input file.
    #[structopt(short, long, parse(from_os_str))]
    pub input: std::path::PathBuf,

    /// Output filename, defaults to `png` output.
    #[structopt(short, long, parse(from_os_str))]
    pub output: Option<std::path::PathBuf>,

    /// Color of the image's foreground, `black`/`b` or `white`/`w`.
    #[structopt(short, long, default_value = "black")]
    pub foreground: String,

    /// Edge thinning algorithm to use, `standard`/`s` or `modified`/`m`.
    #[structopt(short, long, default_value = "modified")]
    pub method: String,

    /// Brightness value below which pixels will become black, ranges from 0.0
    /// to 1.0.
    ///
    /// 0.15 to 0.45 is a reasonable range to start with, some images may
    /// require higher values such as 0.8.
    #[structopt(short, long)]
    pub threshold: Option<f32>,

    /// Run a Sobel edge detection filter on the image before image thinning.
    /// `sobel`/`s` or `sobel4`/`s4` are available options.
    #[structopt(short, long, default_value = "")]
    pub edge: String,

    /// Disables the edge thinning pass, used for generating images with only
    /// thresholding or edge detection performed.
    #[structopt(long)]
    pub no_thin: bool,
}

/// Appends a timestamp to an input filename to be used as the output filename.
fn generate_filename(path: &std::path::Path) -> Result<String, Box<dyn std::error::Error>> {
    let filename = path
        .file_stem()
        .ok_or("No file stem")?
        .to_str()
        .ok_or("Could not convert filename to string")?
        .to_string();
    let now = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH)?;
    let secs = now.as_secs().to_string();
    let millis = format!("{:03}", now.subsec_millis());

    Ok(filename + "-" + &secs + &millis + ".png")
}


src/lib.rs
//! A line thinning library for binary images, including edge detection and
//! threshold functions for preprocessing images into binary images.
//!
//! The goal of thinning is to remove excess pixels from the image until the
//! lines present are one pixel wide, resembling a "skeleton" of the original
//! pattern.
//!
//! The thinning algorithms are based on the papers *Zhang & Suen, 1984* and
//! *Chen & Hsu, 1988*. See [Reference](#reference).
//!
//! ## Usage
//!
//! There are three main workflows for thinning images with this library. The
//! second and third workflows produce binarized images with library functions
//! before thinning the image.
//!
//! The generic [`ForegroundColor`](crate::ForegroundColor) parameter on
//! [`edge_detection::sobel`][sobel], [`edge_detection::sobel4`][sobel4], and
//! [`thin_image_edges`](crate::thin_image_edges) specifies what foreground and
//! background colors the resulting
//! [`thin_image_edges`](crate::thin_image_edges) image will produce. The
//! foreground color is the color of the line to be thinned. A foreground color
//! of white will have a black background and a foreground of black will have a
//! white background. The generic parameters must match when using an edge
//! detection function in combination with the thinning function.
//!
//! [sobel]: crate::edge_detection::sobel
//! [sobel4]: crate::edge_detection::sobel4
//!
//! An example program can be viewed at `/examples/skeletonize.rs`.
//!
//! #### No preprocessing
//!
//! The image is already binarized so the edges can be thinned immediately.
//!
//! ```
//! # fn main() -> Result<(), skeletonize::error::SkeletonizeError> {
//! use skeletonize::{foreground, thin_image_edges, MarkingMethod};
//!
//! # let image_buffer = image::ImageBuffer::from_pixel(1, 1, image::Rgb([255, 255, 255]));
//! # let mut img = image::DynamicImage::ImageRgb8(image_buffer).grayscale();
//! let method = MarkingMethod::Modified;
//!
//! thin_image_edges::<foreground::Black>(&mut img, method, None)?;
//! # Ok(())
//! # }
//! ```
//!
//! If this produces poor results and/or takes a long time to run:
//! - the incorrect foreground color may have been chosen - try using the
//! opposite color, or
//! - the image may not be binary and needs to be thresholded.
//!
//! #### Edge detection
//!
//! Run an edge detection filter on the image and threshold those results before
//! thinning the lines. Note that the foreground color parameters must match on
//! the edge detection function and the thinning function.
//!
//! ```
//! # fn main() -> Result<(), skeletonize::error::SkeletonizeError> {
//! use skeletonize::edge_detection::sobel4;
//! use skeletonize::{foreground, thin_image_edges, MarkingMethod};
//!
//! # let image_buffer = image::ImageBuffer::from_pixel(2, 2, image::Rgb([255, 255, 255]));
//! # let img = image::DynamicImage::ImageRgb8(image_buffer).grayscale();
//! let method = MarkingMethod::Modified;
//! let threshold = Some(0.1);
//!
//! let mut filtered = sobel4::<foreground::White>(&img, threshold)?;
//! thin_image_edges::<foreground::White>(&mut filtered, method, None)?;
//! # Ok(())
//! # }
//! ```
//!
//! #### Thresholding
//!
//! Threshold the image before thinning, e.g., cleaning up a grayscale image.
//!
//! ```
//! # fn main() -> Result<(), skeletonize::error::SkeletonizeError> {
//! use skeletonize::{foreground, thin_image_edges, threshold, MarkingMethod};
//!
//! # let image_buffer = image::ImageBuffer::from_pixel(2, 2, image::Rgb([255, 255, 255]));
//! # let mut img = image::DynamicImage::ImageRgb8(image_buffer).grayscale();
//! let method = MarkingMethod::Modified;
//! let threshold = 0.1;
//!
//! skeletonize::threshold(&mut img, threshold)?;
//! thin_image_edges::<foreground::Black>(&mut img, method, None)?;
//! # Ok(())
//! # }
//! ```
//!
//! ## Reference
//!
//! Zhang, T. Y. & Suen, C. Y. (1984). A fast parallel algorithm for thinning
//! digital patterns. Commun. ACM 27, 3 (March 1984), 236–239.
//! [DOI:10.1145/357994.358023](https://doi.org/10.1145/357994.358023)
//!
//! Chen, Yung-Sheng & Hsu, Wen-Hsing. (1988). A modified fast parallel
//! algorithm for thinning digital patterns. Pattern Recognition Letters. 7.
//! 99-106.
//! [DOI:10.1016/0167-8655(88)90124-9](https://doi.org/10.1016/0167-8655(88)90124-9)
#![warn(missing_docs, rust_2018_idioms, unsafe_code)]

pub mod edge_detection;
pub mod error;
pub mod neighbors;
mod thinning;

use error::{LumaConversionErrorKind, SkeletonizeError};
pub use thinning::thin_image_edges;

/// Represents the color of the foreground or features in a binary image. For
/// example, white text on a black background has a white foreground color and
/// black background color.
pub trait ForegroundColor {
    /// The background color of the image for binarization.
    const BACKGROUND_COLOR: u8;
}

/// Implementations of [`ForegroundColor`](crate::ForegroundColor).
pub mod foreground {
    /// Black foreground color, represented as `0`.
    pub struct Black;

    impl crate::ForegroundColor for Black {
        const BACKGROUND_COLOR: u8 = 255;
    }

    /// White foreground color, represented by `255`.
    pub struct White;

    impl crate::ForegroundColor for White {
        const BACKGROUND_COLOR: u8 = 0;
    }
}

/// Classification of pixels in an image used for edge thinning.
#[derive(Clone, Copy, Debug, PartialEq)]
#[repr(u8)]
pub enum Edge {
    /// The pixel does not contain the foreground color.
    Empty = 0,
    /// The pixel contains the foreground color.
    Filled = 1,
    /// The pixel is not a valid location within the image.
    DoesNotExist,
}

impl Edge {
    /// Convert the edge status into a `u8` representation.
    pub fn to_u8(&self) -> u8 {
        match self {
            Self::Empty | Self::DoesNotExist => 0,
            Self::Filled => 1,
        }
    }
}

/// The algorithm that determines which pixels are removed during the edge
/// thinning process.
///
/// ### Reference
///
/// <span id="standard"></span>Zhang, T. Y. & Suen, C. Y. (1984). A fast
/// parallel algorithm for thinning digital patterns. Commun. ACM 27, 3 (March
/// 1984), 236–239.
/// [DOI:10.1145/357994.358023](https://doi.org/10.1145/357994.358023)
///
/// <span id="modified"></span>Chen, Yung-Sheng & Hsu, Wen-Hsing. (1988). A
/// modified fast parallel algorithm for thinning digital patterns. Pattern
/// Recognition Letters. 7. 99-106.
/// [DOI:10.1016/0167-8655(88)90124-9](https://doi.org/10.1016/0167-8655(88)90124-9)
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum MarkingMethod {
    /// An algorithm based on `Zhang and Suen, 1984`.
    ///
    /// See [MarkingMethod](crate::MarkingMethod#standard) for reference.
    Standard,
    /// An improved and slightly more complex algorithm than `Standard` based on
    /// `Chen and Hsu, 1988`. This algorithm improves on the original's
    /// weaknesses with generally thinner lines and better line connectivity.
    ///
    /// See [MarkingMethod](crate::MarkingMethod#modified) for reference.
    Modified,
}

impl Default for MarkingMethod {
    fn default() -> Self {
        Self::Modified
    }
}

/// Create a binary image where values below `threshold` become black and above
/// become white. `threshold` ranges from 0.0 to 1.0.
pub fn threshold(img: &mut image::DynamicImage, threshold: f32) -> Result<(), SkeletonizeError> {
    for pix in img
        .as_mut_luma8()
        .ok_or(SkeletonizeError::LumaConversion(
            LumaConversionErrorKind::ThresholdMutableLuma,
        ))?
        .iter_mut()
    {
        *pix = if *pix < (threshold * 255.0).round() as u8 {
            0
        } else {
            255
        };
    }

    Ok(())
}
